src/models/necks/segmentation/unet_neck.py

```diff
@@ -0,0 +1,184 @@
+        super().__init__(conv1, conv2)
+
+
+@NECKS.register_class
+class UnetNeck(BaseModel):
+    """Unet is a fully convolution neural network for image semantic segmentation.
+    Paper: https://arxiv.org/pdf/1505.04597.
+    """
+
+    def __init__(self,
+                 encoder_channels: Tuple[int],
+                 decoder_channels: Tuple[int] = (512, 256, 128, 64, 32),
+                 n_blocks: int = 5,
+                 use_batchnorm: bool = True,
+                 use_attention: bool = False,
+                 center: bool = True):
+        """Init Unet
+
+        Args:
+            encoder_channels: Numbers of ``Conv2D`` layer filters.
+            decoder_channels: Numbers of ``Conv2D`` layer filters in decoder blocks.
+            n_blocks: number of stages used in decoder, larger depth - more features are generated.
+                e.g. for depth=3 encoder will generate list of features with following spatial shapes
+                [(H,W), (H/2, W/2), (H/4, W/4), (H/8, W/8)], so in general the deepest feature tensor will have
+                spatial resolution (H/(2^depth), W/(2^depth)]
+            use_batchnorm: If ``True``, ``BatchNormalisation`` applied between every ``Conv2D`` and activation layers.
+            use_attention: If ``True`` will use ``SCSEModule``.
+            center: If ``True`` will use ''CenterBlock''.
+        """
+        super().__init__()
+
+        if n_blocks != len(decoder_channels):
+            raise ValueError(
+                "Model depth is {}, but you provide `decoder_channels` for {} blocks.".format(
+                    n_blocks, len(decoder_channels)
+                )
+            )
+        # channels + input_channels
+        if n_blocks + 1 != len(encoder_channels):
+            raise ValueError(
+                "Model depth is {}, but you provide `encoder_channels` for {} blocks.".format(
+                    n_blocks, len(encoder_channels)
+                )
+            )
+
+        self.n_blocks = n_blocks
+
+        # reverse channels to start from head of encoder
+        encoder_channels = encoder_channels[::-1]
+
+        # computing blocks input and output channels
+        head_channels = encoder_channels[0]
+        in_channels = decoder_channels
+        skip_channels = encoder_channels[1:]
+        out_channels = decoder_channels[1:]
+
+        self.center = CenterBlock(head_channels, head_channels, use_batchnorm=use_batchnorm) if center else None
+
+        blocks = [
+            DecoderBlock(in_ch, skip_ch, out_ch, use_batchnorm=use_batchnorm, use_attention=use_attention)
+            for in_ch, skip_ch, out_ch in zip(in_channels, skip_channels, out_channels)
+        ]
+
+        self.blocks = nn.ModuleList(blocks)
+        self._out_channels = decoder_channels[-1]
+
+    def forward(self, features: List[Tensor]) -> Tensor:
+        """Forward method."""
+        head, *skips = features[::-1]  # reverse channels to start from head of encoder
```

Comment:
Ok skips contain full input image batch, because features return input_image. Is that mean that you process the input image too?